worker/raft/raftbackstop/worker.go

```diff
@@ -0,0 +1,257 @@
+type Config struct {
+	Raft     RaftNode
+	LogStore raft.LogStore
+	Hub      *pubsub.StructuredHub
+	Logger   loggo.Logger
+	Tag      names.Tag
+}
+
+// Validate validates the raft worker configuration.
+func (config Config) Validate() error {
+	if config.Hub == nil {
+		return errors.NotValidf("nil Hub")
+	}
+	if config.Raft == nil {
+		return errors.NotValidf("nil Raft")
+	}
+	if config.LogStore == nil {
+		return errors.NotValidf("nil LogStore")
+	}
+	if config.Tag == nil {
+		return errors.NotValidf("nil Tag")
+	}
+	return nil
+}
+
+// NewWorker returns a worker responsible for recovering the raft
+// cluster when it's been reduced to one server that can't become
+// leader.
+func NewWorker(config Config) (worker.Worker, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+	w := &Worker{
+		config:        config,
+		serverDetails: make(chan apiserver.Details),
+	}
+	// Subscribe to API server address changes.
+	unsubscribe, err := config.Hub.Subscribe(
+		apiserver.DetailsTopic,
+		w.apiserverDetailsChanged,
+	)
+	if err != nil {
+		return nil, errors.Annotate(err, "subscribing to apiserver details")
+	}
+	// Now that we're subscribed, request the current API server details.
+	req := apiserver.DetailsRequest{
+		Requester: "raft-backstop",
+		LocalOnly: true,
+	}
+	if _, err := config.Hub.Publish(apiserver.DetailsRequestTopic, req); err != nil {
+		return nil, errors.Annotate(err, "requesting current apiserver details")
+	}
+
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: func() error {
+			defer unsubscribe()
+			return w.loop()
+		},
+	}); err != nil {
+		unsubscribe()
+		return nil, errors.Trace(err)
+	}
+	return w, nil
+}
+
+type Worker struct {
+	catacomb catacomb.Catacomb
+	config   Config
+
+	serverDetails chan apiserver.Details
+	configUpdated bool
+}
+
+// Kill is part of the worker.Worker interface.
+func (w *Worker) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+// Wait is part of the worker.Worker interface.
+func (w *Worker) Wait() error {
+	return w.catacomb.Wait()
+}
+
+func (w *Worker) loop() error {
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case details := <-w.serverDetails:
+			err := w.maybeRecoverCluster(details)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+}
+
+func (w *Worker) maybeRecoverCluster(details apiserver.Details) error {
+	if w.configUpdated {
+		w.config.Logger.Infof("raft configuration already updated; waiting for raft worker restart")
+		return nil
+	}
+	if len(details.Servers) != 1 {
+		return nil
+	}
+	if _, found := details.Servers[w.config.Tag.Id()]; !found {
+		return nil
+	}
+	if w.config.Raft.State() == raft.Leader {
+		return nil
+	}
+
+	raftServers, err := w.getConfiguration()
+	if err != nil {
+		return errors.Annotate(err, "getting raft configuration")
+	}
+
+	numServers := len(raftServers)
+	localServer := raftServers[raft.ServerID(w.config.Tag.String())]
```

Comment:
Would it be better to key raftServers on tag.Id(). There's no need to use "machine-1" as opposed to just "1". And this would then match details.Servers above, and you could use a machineId var set early in the method.

Reply:
Those server IDs are from the raft configuration - I could use ParseMachineTag to do that in `getConfiguration` but that seemed slightly more awkward than this. 

I guess we could change it so that the raft configuration stored "0" as the id rather than "machine-0" - that would require changing the clusterer and raft worker as well, so I'd do it as a separate change.

Reply:
Yes please, that would be good. Tags really should not be used outside the api layer and the inconsistency here between "0" and "machine-0" makes it worse.

Reply:
Ok, will do.